File: src/ros2_bridge/mappers/geometry_msgs/twist_with_covariance.rs
//! Typed mapper for `geometry_msgs/msg/TwistWithCovariance`.

use crate::ros2_bridge::mapper::TypedTopicMapper;

pub(crate) fn twist_with_covariance_to_bus(msg: ros_env::geometry_msgs::msg::TwistWithCovariance) -> crate::geometry_msgs::msg::v1::TwistWithCovariance {
    crate::geometry_msgs::msg::v1::TwistWithCovariance {
        twist: Some(crate::ros2_bridge::mappers::geometry_msgs::twist::twist_to_bus(msg.twist)),
        covariance: crate::ros2_bridge::mappers::convert::f64_seq(msg.covariance),
    }
}

pub(crate) fn twist_with_covariance_to_ros(bus: crate::geometry_msgs::msg::v1::TwistWithCovariance) -> ros_env::geometry_msgs::msg::TwistWithCovariance {
    ros_env::geometry_msgs::msg::TwistWithCovariance {
        twist: crate::ros2_bridge::mappers::geometry_msgs::twist::twist_to_ros(bus.twist.unwrap_or_default()),
        covariance: crate::ros2_bridge::mappers::convert::FromF64Seq::from_f64_seq(bus.covariance),
    }
}

#[derive(Clone, Copy, Debug, Default)]
pub struct GeometryMsgsTwistWithCovarianceMapper;

impl TypedTopicMapper for GeometryMsgsTwistWithCovarianceMapper {
    type Ros = ros_env::geometry_msgs::msg::TwistWithCovariance;
    type Bus = crate::geometry_msgs::msg::v1::TwistWithCovariance;

    fn type_name(&self) -> &'static str {
        "geometry_msgs/msg/TwistWithCovariance"
    }

    fn ros_to_bus(&self, msg: Self::Ros) -> crate::errors::Result<Self::Bus> {
        Ok(twist_with_covariance_to_bus(msg))
    }

    fn bus_to_ros(&self, msg: Self::Bus) -> crate::errors::Result<Self::Ros> {
        Ok(twist_with_covariance_to_ros(msg))
    }
}

File: bindings/python/robot_bus/ros2_bridge/mappers/geometry_msgs/twist_with_covariance.py
"""Generated mapper for `geometry_msgs/msg/TwistWithCovariance`."""

from __future__ import annotations

from robot_bus.ros2_bridge.mappers import _convert
from robot_bus.ros2_bridge.mappers.geometry_msgs.twist import twist_to_bus, twist_to_ros

_BusMsg = None


def _bus_cls():
    global _BusMsg
    if _BusMsg is None:
        from robot_bus.geometry_msgs.msg.v1 import TwistWithCovariance as BusMsg

        _BusMsg = BusMsg
    return _BusMsg


def twist_with_covariance_to_bus(msg):
    BusMsg = _bus_cls()
    bus = BusMsg()
    bus.twist.CopyFrom(twist_to_bus(msg.twist))
    bus.covariance.extend(list(msg.covariance))
    return bus


def twist_with_covariance_to_ros(bus):
    from geometry_msgs.msg import TwistWithCovariance as RosMsg

    out = RosMsg()
    out.twist = twist_to_ros(bus.twist)
    out.covariance = list(bus.covariance)
    return out


class GeometryMsgsTwistWithCovarianceMapper:
    _ros_type = None

    def ros_msg_type(self):
        cls = type(self)
        if cls._ros_type is None:
            from geometry_msgs.msg import TwistWithCovariance as RosMsg

            cls._ros_type = RosMsg
        return cls._ros_type

    def ros_to_bus(self, msg) -> bytes:
        return twist_with_covariance_to_bus(msg).SerializeToString()

    def bus_to_ros(self, payload: bytes):
        BusMsg = _bus_cls()
        bus = BusMsg()
        bus.ParseFromString(payload)
        return twist_with_covariance_to_ros(bus)
